src/Mod/Sketcher/Gui/TaskSketcherElements.cpp

```diff
@@ -449,147 +609,129 @@ void TaskSketcherElements::onSelectionChanged(const Gui::SelectionChanges& msg)
     }
 }
 
+void TaskSketcherElements::on_listWidgetElements_itemPressed(QListWidgetItem* itf) {
+    //We use itemPressed instead of previously used ItemSelectionChanged because if user click on already selected item, ItemSelectionChanged didn't trigger.
+    if (itf == nullptr) { return; }
 
-void TaskSketcherElements::on_listWidgetElements_itemSelectionChanged(void)
-{
-    ui->listWidgetElements->blockSignals(true);
-
-
-    // selection changed because we acted on the current entered item
-    // we can not do this with ItemPressed because that signal is triggered after this one
-    int element=ui->comboBoxElementFilter->currentIndex();
-
-    ElementItem * itf;
-
-    if(focusItemIndex>-1 && focusItemIndex<ui->listWidgetElements->count())
-      itf=static_cast<ElementItem*>(ui->listWidgetElements->item(focusItemIndex));
-    else
-      itf=nullptr;
-
-    bool multipleselection=true; // ctrl type of selection in listWidget
-    bool multipleconsecutiveselection=false; // shift type of selection in listWidget
-
-    if (!inhibitSelectionUpdate) {
-        if (itf) {
-            switch(element){
-            case 0:
-                itf->isLineSelected=!itf->isLineSelected;
-                break;
-            case 1:
-                itf->isStartingPointSelected=!itf->isStartingPointSelected;
-                break;
-            case 2:
-                itf->isEndPointSelected=!itf->isEndPointSelected;
-                break;
-            case 3:
-                itf->isMidPointSelected=!itf->isMidPointSelected;
-                break;
-            }
-        }
+    ElementData* itfData = qvariant_cast<ElementData*>(itf->data(Qt::UserRole));
+    bool rightClickOnSelected = itfData->rightClicked && (itfData->isLineSelected || itfData->isStartingPointSelected || itfData->isEndPointSelected || itfData->isMidPointSelected);
+    itfData->rightClicked = false;
+    if (rightClickOnSelected) { return; } //if user right clicked on a selected item, change nothing.
 
-        if (QApplication::keyboardModifiers()==Qt::ControlModifier)// multiple ctrl selection?
-            multipleselection=true;
-        else
-            multipleselection=false;
+    ui->listWidgetElements->blockSignals(true);
 
-        if (QApplication::keyboardModifiers()==Qt::ShiftModifier)// multiple shift selection?
-            multipleconsecutiveselection=true;
-        else
-            multipleconsecutiveselection=false;
+    bool multipleselection = false;
+    bool multipleconsecutiveselection = false;
+    if (QApplication::keyboardModifiers() == Qt::ControlModifier)
+        multipleselection = true;
+    if (QApplication::keyboardModifiers() == Qt::ShiftModifier)
+        multipleconsecutiveselection = true;

```

Comment:
This isn't very robust.
Why can't you still use `itemSelectionChange`. You already have a mouse press event dealing with subitems selection.

Reply:
Why do you say it's not robust? I could have keep itemSelectionChange, but it still required to implement itemPressed to catch the press item, then check if itemSelectionChange wasn't already triggered. So it added some complexity.
So I simplified this way. And I can't find any issue with it.
Please state if you have any specific issue.

Reply:
It's not robust because it may virtually happen that user release the modifier key between click and function call.

Reply:
What? But why wouldn't it be true with itemSelectionChange event?

Reply:
I guess from there you can directly see if new selection holds several items.